docs/connectors/completion-tokens.md

```diff
@@ -0,0 +1,152 @@
+# Synchronous Processing with Completion Tokens
+
+Completion tokens enable synchronous processing on top of Feldera’s asynchronous pipelines, allowing
+users to precisely track when a specific set of inputs has been fully processed by the pipeline.
+
+## Background
+
+Feldera pipelines process data **asynchronously**. While the incremental SQL engine is evaluating
+earlier inputs, input connectors continue ingesting data into internal buffers, and output connectors
+write prior results to external sinks.
+
+![Architecture of a Feldera Pipeline](../pipelines/pipeline_architecture.png)
+
+However, the SQL engine itself is **synchronous** and **strongly consistent**. It processes inputs in
+discrete transactions—each transaction: ingests a set of records from input connectors, processes them
+to completion and outputs **all** resulting changes to the output views before moving on to the next
+transaction.
+
+This enables Feldera to determine precisely when a specific set of inputs has been fully processed,
+from ingest all the way to writing results to data sinks. Completion tokens expose this capability via
+a simple API.
+
+## Completion tokens
```

Comment:
this calls for a diagram where you show multiple tokens "flowing" through the pipeline and where the completed ones made it to the output.